### ISSUE TITLE
Add support for Python 3.8

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#4311]
This will be more effort than for Python 3.7 since the Python initialization sequence has been changes and the bootloader needs quite some changes.

Anybody going to [fund this development](https://www.pyinstaller.org/funding.html)? Otherwise it might be a long time until this will happen.

**Comment by lkollar** [comment]
Thanks to @vstinner and @pablogsal we now have #4440 and #4441 which will make PyInstaller work on Python 3.8. They discovered a bug in 3.8 which Victor has already fixed and should be in rc1:  https://bugs.python.org/issue38234. This means that with the PRs applied PyInstaller should work with the upcoming Python 3.8rc1 version.

Once rc1 is out we should add it to Travis, but we will need at least #4440 landed first.

**Comment by jay0lee** [comment]
FYI I attempted a build with 3.8rc1 and got this crash:

$ python -OO -m PyInstaller --clean --noupx --strip -F --distpath=gam linux-gam.spec
321 INFO: PyInstaller: 3.5
321 INFO: Python: 3.8.0rc1
327 INFO: Platform: Linux-4.19.0-6-amd64-x86_64-with-glibc2.2.5
330 INFO: UPX is not available.
331 INFO: Removing temporary files and cleaning cache in /home/jay/.cache/pyinstaller
349 INFO: Extending PYTHONPATH with paths
['/home/jay/GAM/src']
349 INFO: checking Analysis
349 INFO: Building Analysis because Analysis-00.toc is non existent
350 INFO: Initializing module dependency graph...
353 INFO: Initializing module graph hooks...
357 INFO: Analyzing base_library.zip ...
13757 INFO: running Analysis Analysis-00.toc
13793 INFO: Caching module hooks...
13807 INFO: Analyzing gam.py
18564 INFO: Processing pre-safe import module hook   six.moves
20456 INFO: Processing pre-find module path hook   distutils
23033 INFO: Processing pre-safe import module hook   urllib3.packages.six.moves
29042 INFO: Loading module hooks...
29043 INFO: Loading module hook "hook-httplib2.py"...
29045 INFO: Loading module hook "hook-certifi.py"...
29048 INFO: Loading module hook "hook-pkg_resources.py"...
29945 INFO: Processing pre-safe import module hook   win32com
30484 INFO: Loading module hook "hook-distutils.py"...
30491 INFO: Loading module hook "hook-pydoc.py"...
30493 INFO: Loading module hook "hook-xml.py"...
30643 INFO: Loading module hook "hook-sysconfig.py"...
30668 INFO: Loading module hook "hook-encodings.py"...
30898 INFO: Looking for ctypes DLLs
31011 WARNING: library user32 required via ctypes not found
32128 INFO: Analyzing run-time hooks ...
32156 INFO: Including run-time hook 'pyi_rth_certifi.py'
32159 INFO: Including run-time hook 'pyi_rth_pkgres.py'
32162 INFO: Including run-time hook 'pyi_rth_multiprocessing.py'
32207 INFO: Looking for dynamic libraries
33133 INFO: Looking for eggs
33133 INFO: Using Python library /home/jay/python/lib/libpython3.8.so.1.0
33169 INFO: Warnings written to /home/jay/GAM/src/build/linux-gam/warn-linux-gam.txt
33388 INFO: Graph cross-reference written to /home/jay/GAM/src/build/linux-gam/xref-linux-gam.html
33530 INFO: checking PYZ
33530 INFO: Building PYZ because PYZ-00.toc is non existent
33530 INFO: Building PYZ (ZlibArchive) /home/jay/GAM/src/build/linux-gam/PYZ-00.pyz
Traceback (most recent call last):
  File "/home/jay/python/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/jay/python/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/__main__.py", line 118, in <module>
    run()
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/__main__.py", line 111, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/__main__.py", line 63, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/build_main.py", line 844, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/build_main.py", line 791, in build
    exec(code, spec_namespace)
  File "linux-gam.spec", line 18, in <module>
    pyz = PYZ(a.pure)
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/api.py", line 98, in __init__
    self.__postinit__()
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/datastruct.py", line 158, in __postinit__
    self.assemble()
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/api.py", line 128, in assemble
    self.code_dict = {
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/api.py", line 129, in <dictcomp>
    key: strip_paths_in_code(code)
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/utils.py", line 652, in strip_paths_in_code
    consts = tuple(
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/utils.py", line 653, in <genexpr>
    strip_paths_in_code(const_co, new_filename)
  File "/home/jay/python/lib/python3.8/site-packages/PyInstaller/building/utils.py", line 660, in strip_paths_in_code
    return code_func(co.co_argcount, co.co_kwonlyargcount, co.co_nlocals, co.co_stacksize,
TypeError: an integer is required (got type bytes)

**Comment by pablogsal** [comment]
This is fixed in #4440

**Comment by arossert** [comment]
Is there any scheduled release for an official Python 3.8 support?

**Comment by jakirkham** [comment]
> Is there any scheduled release for an official Python 3.8 support?

Have raised issue ( https://github.com/pyinstaller/pyinstaller/issues/4515 ) to get some insight on this.

**Comment by htgoebel** [comment]
No yet implemented.

**Comment by htgoebel** [comment]
@lkollar This is all which is required for python 3.8? Cool. I was afraid there will be more work to be done. Will review this soon.

**Comment by lkollar** [comment]
Python 3.8 should be added to the CI build matrix as well, but I can't think of anything else.

**Comment by anthonyr013** [comment]
support python 3.8,0 ;V

**Comment by The-Compiler** [comment]
@anthonyr013 Comments of this kind aren't helpful at all. All you do is spamming at least 225 people (plus whoever is subscribed to this issue) with a notification with zero content. If you want to see things being done faster, consider contributing, either time/knowledge or financially.

**Comment by htgoebel** [comment]
Implementing this still needs funding.

**Comment by Legorooj** [comment]
@htgoebel out of curiosity what would the figures be?

**Comment by LefterisJP** [comment]
@htgoebel Would you accept cryptocurrencies for funding? I completely understand and appreciate the problem that opensource maintainers face with funding.

Incidentally right now and until the 21st of this month there is a [quadratic funding](https://vitalik.ca/general/2019/12/07/quadratic.html) round at [Gitcoin](https://gitcoin.co/grants/), an ethereum based platform for funding open source projects.

Each donation gets matched by a big pool of money put in by third parties. In this round it's the Ethereum foundation that has put up that money.

The projects do not need to strictly be cryptocurrency related. There is a grant for [python urllib3](https://gitcoin.co/grants/65/urllib3) there. Also you don't need to be exposed to cryptocurrency price fluctuations. Most of the donations are made in a stablecoin called DAI which is pegged to the dollar. You can specify that you only accept DAI when creating the grant.

I know many projects, including one I maintain myself, that are part of the ethereum circle and use pyinstaller. I am sure that people would donate and their donations get matched by the pool so as a result you could get a good amount of funding from there.

I can promise you I would be one of the people to donate if you offered such an option. If you decide to try it and need guidance with anything related to ethereum or how to create a gitcoin grant please contact me. Best way is via email: lefteris [at] rotki [dot] com

**Comment by Legorooj** [comment]
@LefterisJP the Bitcoin address is on the donate page at PyInstaller.org

**Comment by htgoebel** [comment]
@LefterisJP My landlord and the bakery both do not accept cryptocurrencies. So: No, I'm not accepting cryptocurrencies for funding.

**Comment by Legorooj** [comment]
@htgoebel did you get my email? Legorooj@protonmail.com

**Comment by wyz23x2** [comment]
#4835

**Comment by alexchandel** [comment]
What changes are necessary for 3.8 support PR?  I count:

[ ] #4868 (spawn support), which fixes #4865
[ ] #4536 (imp DeprecationWarning)
[ ] #4406 (RecursionError) possibly
[ ] #4659 (PySide) possibly
[ ] #4441 (Fix pyi_locale_char2wchar), although @htgoebel indicates he has a private fix

What other enumerable remain?

**Comment by alexchandel** [comment]
Some optional bootloader changes from PEP 587's new initialization may include:

Path configuration:

* Use `PyConfig->program_name` instead of [PI_Py_SetProgramName(progname_w)](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L409)
* Use `PyConfig->home` instead of [Py_SetPythonHome(pyhome_w)](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L420)
* Use `PyConfig->pythonpath_env` instead of [Py_SetPath(pypath_w)](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L453)

Global variables:

* Use `PyConfig->site_import` instead of [Py_NoSiteFlag = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L196)
* Use `PyConfig->pathconfig_warnings` instead of [Py_FrozenFlag = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L198)
* Use `PyConfig->write_bytecode` instead of [Py_DontWriteBytecodeFlag = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L200)
* Use `PyConfig->user_site_directory` instead of [Py_NoUserSiteDirectory = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L202) (done by `PyConfig->isolated`)
* Use `PyConfig->use_environment` instead of [Py_IgnoreEnvironmentFlag = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L204) (done by `PyConfig->isolated`)
* Use `PyConfig->verbose` instead of [Py_VerboseFlag = 0](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L206)
* Use `PyConfig->optimization_level` instead of [Py_OptimizeFlag = 1](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L236)
* Use `PyConfig->buffered_stdio` instead of [the unbuffering logic here](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L242)

Arguments and `sys.path`:

* Use `Py_InitializeFromConfig` instead of [Py_Initialize](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L474), if any `PyConfig` are used
* Use `PyConfig->module_search_paths` or perhaps `PyPreConfig->isolated` and `PyConfig->isolated` instead of [PySys_SetPath(pypath_w)](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L488), if the effect is desired before `Py_Initialize`.  `isolated` may also simplify the isolation steps [mentioned here](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L482)
* Use `PyConfig->argv` instead of [PySys_SetArgvEx(status->argc, wargv, 0)](https://github.com/pyinstaller/pyinstaller/blob/1409fae6878a432e869f347fa8691700ce663288/bootloader/src/pyi_pythonlib.c#L343), if the effect is desired before `Py_Initialize`

However, all of these changes are optional.  Python 3.8 doesn't require them.  Is there anything in the bootloader that actually **must** be changed before a release?

**Comment by Legorooj** [comment]
@alexchandel the imp DeprecationWarning and the RecursionError can be taken off the list; I've got those covered in the entire modulegraph replacement I'm doing for the 4.1 release.

Otherwise I can't think of anything else off the top of my head that would need doing, though the person to ask is @htgoebel. I don't know how much of the work he's already done either, so you need to ask him that as well.

**Comment by alexchandel** [comment]
@Legorooj Ok. Sounds like it isn't a blocker either. Let's merge #4868, and merge @htgoebel's "private fix" for #4441, and release a Python 3.8 version.

**Comment by htgoebel** [comment]
Good news: I got some funding for developing this. Expect a release supporting Python 3.8 really soon now.

**Comment by htgoebel** [comment]
Unfortunately I'm facing very curious bugs, none of which i can reproduce locally:
* On appveyor, [test_gevent keeps failing due to some issue in hook-pyside](https://ci.appveyor.com/project/htgoebel/pyinstaller/builds/34893829/job/3rs7cm9v9hfc9pl1)
* Same tests [pass on gtihub actions](https://github.com/htgoebel/pyinstaller/actions/runs/227721741), but github actions [is unrelyable](https://github.com/htgoebel/pyinstaller/actions/runs/227765902)
* On Travis [test_multiprocess_spawn_pool failes](https://travis-ci.org/github/htgoebel/pyinstaller/jobs/721610618#L2316), even with spawn-support enabled.

Any help on debugging is appreciated.

**Comment by Jack2104** [comment]
Is this fixed now?

**Comment by judge2020** [comment]
@Jack2104 the PR was merged in, so it should be fairly stable, but there hasn't been a release with that PR yet. If you install from git (see below), you should have the PR and the latest changes, although there might be more bugs than in a release branch.

Bare pip: `pip install https://github.com/pyinstaller/pyinstaller/archive/develop.tar.gz` 

pipenv/pipfile:

```
pyinstaller = {file = "https://github.com/pyinstaller/pyinstaller/archive/develop.tar.gz"}
```

**Comment by mram1000** [comment]
This still does't work for me :

Bare pip: `pip install https://github.com/pyinstaller/pyinstaller/archive/develop.tar.gz`

Is there any other option?

**Comment by judge2020** [comment]
@mram1000 If you ran that (and have confirmed you're actually running the new pyinstaller version), you should open a new issue or search for an existing issue that matches the problem you're having. Be sure to include as much debug info as you have in your report.

**Comment by Legorooj** [comment]
@Jack2104 there should be another release for 3.8 support soon - we're just waiting on Hartmut to actually do it.

**Comment by mram1000** [comment]
> @mram1000 If you ran that (and have confirmed you're actually running the new pyinstaller version), you should open a new issue or search for an existing issue that matches the problem you're having. Be sure to include as much debug info as you can in your report.

Hi here is my error output - this time I downloaded py3.7 but still same error - so not sure what it might be? Appreciate any help please! Thanks in advance!

```
 c:\Python\Python37\Scripts>pip install pyinstaller -
-proxy http://user:pass@proxy-us.intel.com:912
Collecting pyinstaller
  Using cached pyinstaller-4.0.tar.gz (3.5 MB)
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: 'c:\python\python37\python.exe' 'c:\python\python37\lib\site-packages\pip' install --ignore-installed --no-user --prefix 'C:\Users\mram\AppData\Local\Temp\pip-build-env-ez_klbyj\overlay' --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel
       cwd: None
  Complete output (9 lines):
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000020C87AE8F88>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/setuptools/
  WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000020C87AF2788>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/setuptools/
  WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000020C87AFD988>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/setuptools/
  WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000020C87AFDD88>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/setuptools/
  WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x0000020C87B19208>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/setuptools/
  ERROR: Could not find a version that satisfies the requirement setuptools>=40.8.0 (from versions: none)
  ERROR: No matching distribution found for setuptools>=40.8.0
  WARNING: You are using pip version 20.1.1; however, version 20.2.3 is available.
  You should consider upgrading via the 'c:\python\python37\python.exe -m pip install --upgrade pip' command.
  ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\python\python37\python.exe' 'c:\python\python37\lib\site-packages\pip' install --ignore-installed --no-user --prefix 'C:\Users\mram\AppData\Local\Temp\pip-build-env-ez_klbyj\overlay' --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel Check the logs for full command output.
```

**Comment by campanelli-resonon** [comment]
@mram1000 Can you successfully execute `pip install --upgrade setuptools`?

**Comment by mram1000** [comment]
> @mram1000 Can you successfully execute `pip install --upgrade setuptools`?

Not without specifying the proxy! 

OK so after disconnecting from my corporate overlords, I was able to install 3.7 version - thank you for your help!!

**Comment by htgoebel** [comment]
@mram1000 STOP SPAMMING THIS CLOSED ISSUE! You have been asked to open a new one!

**Comment by htgoebel** [comment]
Delete conversation with a user not willed to read our wiki or our manual, neither following our advice.

**Comment by TechmarkDavid** [comment]
https://github.com/pyinstaller/pyinstaller/issues/5281  ZeroMQ is broken with PyInstaller on 3.8

**Comment by Legorooj** [comment]
@TechmarkDavid please open a new issue.
